vm/starnix/starnix.go

```diff
@@ -0,0 +1,374 @@
+	}
+	inst.fuchsiaVmStarted = true
+	return nil
+}
+
+func (inst *instance) startFuchsiaLogs() (*exec.Cmd, error) {
+	cmd := osutil.Command("ffx", "--target", inst.name, "log")
+	cmd.Dir = inst.fuchsiaDirectory
+	cmd.Stdout = inst.wpipe
+	cmd.Stderr = inst.wpipe
+	inst.merger.Add("fuchsia", inst.rpipe)
+	return cmd, cmd.Start()
+}
+
+func (inst *instance) startAdbServerAndConnection(timeout time.Duration) error {
+	cmd := osutil.Command("ffx", "--target", inst.name, "starnix", "adb", "-p", fmt.Sprintf("%d", inst.port))
+	cmd.Dir = inst.fuchsiaDirectory
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	if inst.debug {
+		log.Logf(0, fmt.Sprintf("The adb bridge is listening on 127.0.0.1:%d", inst.port))
+	}
+	inst.adb = cmd
+	return inst.connectToAdb(timeout)
+}
+
+func (inst *instance) connectToAdb(timeout time.Duration) error {
+	startTime := time.Now()
+	for {
+		select {
+		case <-time.After(3 * time.Second):
+		}
+		if inst.debug {
+			log.Logf(0, "Attempting to connect to ADB")
+		}
+		cmd := osutil.Command("adb", "connect", fmt.Sprintf("127.0.0.1:%d", inst.port))
+		connect_output, err := cmd.Output()
+		if err == nil && strings.HasPrefix(string(connect_output), "connected to") {
+			return nil
+		}
+		cmd = osutil.Command("adb", "disconnect", fmt.Sprintf("127.0.0.1:%d", inst.port))
+		cmd.Run()
+		if inst.debug {
+			log.Logf(0, "adb connect failed")
+		}
+		if time.Since(startTime) > timeout {
+			return fmt.Errorf("Can't connect to ADB server")
+		}
+	}
+}
+
+// Script for telling syz-fuzzer how to connect to syz-executor
+func (inst *instance) createAdbScript() error {
+	adbScript := fmt.Sprintf(
+		`#!/bin/bash
+		adb_port=$1
+		fuzzer_args=${@:2}
+		adb -s 127.0.0.1:$adb_port shell "cd %s; ./syz-executor $fuzzer_args"`, inst.targetDir())
+	return os.WriteFile(inst.executor, []byte(adbScript), 0777)
+}
+
+func (inst *instance) boot() error {
+	inst.port = vmimpl.UnusedTCPPort()
+	// Start output merger.
+	var tee io.Writer
+	if inst.debug {
+		tee = os.Stdout
+	}
+	inst.merger = vmimpl.NewOutputMerger(tee)
+
+	inst.terminateFuchsiaVm()
+
+	if err := inst.startFuchsiaVm(); err != nil {
+		return err
+	}
+
+	if err := inst.startAdbServerAndConnection(1 * time.Minute * inst.timeouts.Scale); err != nil {
+		return err
+	}
+
+	if err := inst.createAdbScript(); err != nil {
+		return err
+	}
+
+	cmd, err := inst.startFuchsiaLogs()
+	if err != nil {
+		return err
+	}
+	inst.fuchsiaLogs = cmd
+	if inst.debug {
+		log.Logf(0, "%s booted successfully, inst.port = %d", inst.name, inst.port)
+	}
+	return nil
+}
+
+// Runs a command inside the fuchsia directory and redirects the output to the
+// instance output merger
+func (inst *instance) runCommand(args ...string) (*exec.Cmd, error) {
+	rpipe, wpipe, err := osutil.LongPipe()
+	if err != nil {
+		return nil, err
+	}
+	cmd := osutil.Command(args[0], args[1:]...)
+	cmd.Dir = inst.fuchsiaDirectory
+	cmd.Stdout = wpipe
+	cmd.Stderr = wpipe
+	inst.merger.Add(args[0], rpipe)
+	err = cmd.Run()
```

Comment:
Using pipes, etc only makes sense if you use cmd.Start and it runs in background.
Since we run it to completion and wait, use pkg/osutil.RunCmd.
Does the output needs to go into inst.merger? Why?

Reply:
It was mainly for debugging, changing it to use osutil.RunCmd and throw the output to stdout on debug